source/grpc_async_client_impl.cc

```diff
@@ -0,0 +1,136 @@
+#include "utils/grpc_status.h"
+
+namespace cpp2sky {
+
+void* toTag(TaggedStream* stream) { return reinterpret_cast<void*>(stream); }
+TaggedStream* deTag(void* stream) { return static_cast<TaggedStream*>(stream); }
+
+TracerStubImpl::TracerStubImpl(std::shared_ptr<grpc::Channel> channel)
+    : stub_(TraceSegmentReportService::NewStub(channel)) {}
+
+std::unique_ptr<grpc::ClientAsyncWriter<TracerRequestType>>
+TracerStubImpl::createWriter(grpc::ClientContext* ctx,
+                             TracerResponseType* response,
+                             grpc::CompletionQueue* cq, void* tag) {
+  return stub_->Asynccollect(ctx, response, cq, tag);
+}
+
+GrpcAsyncSegmentReporterClient::GrpcAsyncSegmentReporterClient(
+    grpc::CompletionQueue* cq,
+    AsyncStreamFactory<TracerRequestType, TracerResponseType>& factory,
+    std::shared_ptr<grpc::ChannelCredentials> cred, std::string address)
+    : address_(address), factory_(factory), cq_(cq) {
+  stub_ = std::make_unique<TracerStubImpl>(grpc::CreateChannel(address, cred));
+  stream_ = factory_.create(this);
+  stream_->startStream();
+}
+
+void GrpcAsyncSegmentReporterClient::sendMessage(Message& message) {
+  GPR_ASSERT(stream_ != nullptr);
+  stream_->sendMessage(message);
+}
+
+std::unique_ptr<grpc::ClientAsyncWriter<TracerRequestType>>
+GrpcAsyncSegmentReporterClient::createWriter(grpc::ClientContext* ctx,
+                                             TracerResponseType* response,
+                                             void* tag) {
+  return stub_->createWriter(ctx, response, cq_, tag);
+}
+
+GrpcAsyncSegmentReporterStream::GrpcAsyncSegmentReporterStream(
+    AsyncClient<TracerRequestType, TracerResponseType>* client)
+    : client_(client) {}
+
+GrpcAsyncSegmentReporterStream::~GrpcAsyncSegmentReporterStream() {
+  ctx_.TryCancel();
+  request_writer_->Finish(&status_, toTag(&finish_));
+}
+
+bool GrpcAsyncSegmentReporterStream::startStream() {
+  request_writer_.reset();
+
+  // Ensure pending RPC will complete if connection to the server is not
+  // established first because of like server is not ready. This will queue
+  // pending RPCs and when connection has established, Connected tag will be
+  // sent to CompletionQueue.
+  ctx_.set_wait_for_ready(true);
+  request_writer_ =
+      client_->createWriter(&ctx_, &commands_, toTag(&connected_));
+  return true;
+}
+
+void GrpcAsyncSegmentReporterStream::sendMessage(Message& message) {
+  pending_messages_.emplace(message);
+  clearPendingMessages();
+}
+
+bool GrpcAsyncSegmentReporterStream::clearPendingMessages() {
+  if (state_ != Operation::Idle || pending_messages_.empty()) {
+    return false;
+  }
+  auto message = pending_messages_.back();
+  pending_messages_.pop();
+  TracerRequestType obj;
+  obj.CopyFrom(message.get());
+  request_writer_->Write(obj, toTag(&write_done_));
+  return true;
+}
+
+bool GrpcAsyncSegmentReporterStream::handleOperation(Operation incoming_op) {
+  state_ = incoming_op;
+  while (true) {
```

Comment:
Isn't this a busy loop when state_ is Connected or WriteDone? Did I miss anything?

Reply:
the state will walk `WriteDone` -> `Idle` or `Connected` -> `Idle`. But it will get out from this loop when the state is `Idle`. So we won't face busy loop.

Reply:
I would prefer not use `while (true)` and `switch` together since it's really hard to read. Instead:


```
if (state_ == Connected) { state_ = Idle; ...}
if (state_ == WriteDone) { state_ = Idle; ...}
if (state_ == Idle } {...; return true;}
if (state_ == Finished(...; return false;}
throw ...
```
will be much clearer.

Reply:
done.

Reply:
You shouldn't need `while(true)` either, if you `s/else if/if/`